backend/httpclient/datasource_metrics_middleware.go

```diff
@@ -0,0 +1,153 @@
+			Name:      "datasource_request_in_flight",
+			Help:      "A gauge of outgoing external data source requests currently being sent by Grafana",
+		},
+		[]string{"datasource", "datasource_type", "secure_socks_ds_proxy_enabled"},
+	)
+)
+
+// SanitizeLabelName removes all invalid chars from the label name.
+// If the label name is empty or contains only invalid chars, it
+// will return an error.
+func SanitizeLabelName(name string) (string, error) {
+	if len(name) == 0 {
+		return "", errors.New("label name cannot be empty")
+	}
+
+	out := strings.Builder{}
+	for i, b := range name {
+		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
+			out.WriteRune(b)
+		} else if b == ' ' {
+			out.WriteRune('_')
+		}
+	}
+
+	if out.Len() == 0 {
+		return "", fmt.Errorf("label name only contains invalid chars: %q", name)
+	}
+
+	return out.String(), nil
+}
+
+const DataSourceMetricsMiddlewareName = "datasource_metrics"
+
+var executeMiddlewareFunc = executeMiddleware
+
+func DataSourceMetricsMiddleware() Middleware {
+	return NamedMiddlewareFunc(DataSourceMetricsMiddlewareName, func(opts Options, next http.RoundTripper) http.RoundTripper {
+		if opts.Labels == nil {
+			return next
+		}
+
+		datasourceName, exists := opts.Labels["datasource_name"]
+		if !exists {
+			return next
+		}
+
+		datasourceLabelName, err := SanitizeLabelName(datasourceName)
+		// if the datasource named cannot be turned into a prometheus
+		// label we will skip instrumenting these metrics.
+		if err != nil {
+			return next
+		}
+
+		datasourceType, exists := opts.Labels["datasource_type"]
+		if !exists {
+			return next
+		}
+		datasourceLabelType, err := SanitizeLabelName(datasourceType)
+		// if the datasource type cannot be turned into a prometheus
+		// label we will skip instrumenting these metrics.
+		if err != nil {
+			return next
+		}
+
+		labels := prometheus.Labels{
+			"datasource":                    datasourceLabelName,
+			"datasource_type":               datasourceLabelType,
+			"secure_socks_ds_proxy_enabled": strconv.FormatBool(opts.ProxyOptions != nil && opts.ProxyOptions.Enabled),
+		}
+
+		return executeMiddlewareFunc(next, labels)
+	})
+}
+
+func executeMiddleware(next http.RoundTripper, labels prometheus.Labels) http.RoundTripper {
+	return RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
+		requestCounter := datasourceRequestCounter.MustCurryWith(labels)
+		requestHistogram := datasourceRequestHistogram.MustCurryWith(labels)
+		requestInFlight := datasourceRequestsInFlight.With(labels)
+		responseSizeHistogram := datasourceResponseHistogram.With(labels)
+
+		res, err := promhttp.InstrumentRoundTripperDuration(requestHistogram,
+			promhttp.InstrumentRoundTripperCounter(requestCounter,
+				promhttp.InstrumentRoundTripperInFlight(requestInFlight, next))).
+			RoundTrip(r)
+		if err != nil {
+			return nil, err
+		}
+
+		if res != nil && res.StatusCode != http.StatusSwitchingProtocols {
```

Comment:
Maybe an improvement for down the road but I wonder how ubiquitous `context-length` is as a http header. Potentially, we could short circuit this if we have the information already in the response headers

Reply:
We tried that a first/couple years ago, but is not reliably returned/set in all responses